### PR TITLE
[exporter/loki] Log the first part of the http body on failed pushes to loki

### PR DESCRIPTION
**Pull request by gregoryfranklin**
The body includes important information about the reason for the error.

This is based on the promtail implementation
https://github.com/grafana/loki/blob/bd1fef08a8ea99e73a30c4739cfeb931353700e6/clients/pkg/promtail/client/client.go#L390

### CODE REVIEW

File: exporter/lokiexporter/exporter.go [hunk 1]
```diff
@@ -93,11 +97,18 @@ func (l *lokiExporter) pushLogData(ctx context.Context, ld pdata.Logs) error {
...
+	}()
 
 	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
-		err = fmt.Errorf("HTTP %d %q", resp.StatusCode, http.StatusText(resp.StatusCode))
+		scanner := bufio.NewScanner(io.LimitReader(resp.Body, maxErrMsgLen))
+		line := ""
+		if scanner.Scan() {
+			line = scanner.Text()
+		}
+		err = fmt.Errorf("HTTP %d %q: %s", resp.StatusCode, http.StatusText(resp.StatusCode), line)
```
Comment (jpkrohling) [inline exporter/lokiexporter/exporter.go:111]: Would it make sense to wrap this in a `sync.Once`, or a [rate limiter](https://pkg.go.dev/golang.org/x/time/rate) (1 per minute?), so that we don't flood the collector logs while still providing useful debugging information to users?

Comment (gregoryfranklin) [reply exporter/lokiexporter/exporter.go:111]: It only shows the first 1024 chars of the error message to limit the extent to which the collector logs are flooded.  Personally, I think this is sufficient.  It would probably be fine if this limit were this were smaller - most of the important information about the reason for the error is at the beginning of the message.  So, I'd be happy to set maxErrMsgLen to a smaller number.

Currently, you do get a log line on every failed push, its just that it only says something like "HTTP 429".  So, rate limiting wouldn't change the number of messages logged, only the length of the message (unless we were to change that as well).

If we rate limit the messages, would we need to check the signature of the log message so that the message is only suppressed if it is the same as previous messages?  This seems like it would be quite complicated (requires keeping a lot of state to track which messages have been shown).
Therefore, if we were to do rate limiting, it would just be for whether or not to show the body.
eg
```
limiter := rate.NewLimiter(rate.Every(time.Minute), 0)

line := ""
if l.limiter.Allow() {
  line = scanner.Text()
} else {
  line = "message suppressed (rate limiting)"
}
```
If you were to get different errors, you would only see the reason for the first error.

A alternative approach might be to have different truncation lengths.  So the first error gets truncated to 1024 chars and subsequent errors get truncated to 50 chars?
eg
```
const (
        maxErrMsgLenShort = 50
        maxErrMsgLenLong = 1024
)

limiter := rate.NewLimiter(rate.Every(time.Minute), 0)

if l.limiter.Allow() {
  scanner := bufio.NewScanner(io.LimitReader(resp.Body, maxErrMsgLenLong))
} else {
  scanner := bufio.NewScanner(io.LimitReader(resp.Body, maxErrMsgLenShort))
}
```

In summary, I think the current approach of truncating the message to 1024 chars is fine and I could reduce this limit if there are concerns.
Using a rate limiter to truncate to 1024 chars for the first error and 50 chars for subsequent errors would also be fine from my perspective, but seems unnecessary for what is supposed to only be an error condition.
Or we could use the rate limiter to suppress the error message entirely (including the "HTTP 429" part).  But personally, I would have thought that if somebody needed this, it should be done by whatever processes the collector logs rather than in the collector itself.

Comment (jpkrohling) [reply exporter/lokiexporter/exporter.go:111]: Alright, I think it's fine. I was concerned that the logs would get flooded in case of an outage of Loki, as every request would fail for a given window of time, but perhaps zap would do rate limiting itself as well.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Sorry, but it looks like the changelog now has a conflict.

**Review by jpkrohling — APPROVED** [review]
